### PR TITLE
[DataGrid] Add customisable `aria-label`, `aria-labelledby` field

### LINKED ISSUE #1290: [RFC] Remove data grid aria-label
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

```diff
diff --git a/packages/grid/_modules_/grid/components/containers/GridRoot.tsx b/packages/grid/_modules_/grid/components/containers/GridRoot.tsx
index 6ebca39f..c65f26ca 100644
--- a/packages/grid/_modules_/grid/components/containers/GridRoot.tsx
+++ b/packages/grid/_modules_/grid/components/containers/GridRoot.tsx
@@ -28,7 +28,6 @@ export const GridRoot = React.forwardRef<HTMLDivElement, GridRootProps>(function
       aria-colcount={visibleColumnsLength}
       aria-rowcount={gridState.rows.totalRowCount}
       tabIndex={0}
-      aria-label={apiRef!.current.getLocaleText('rootGridLabel')}
       aria-multiselectable={!gridState.options.disableMultipleSelection}
       {...other}
     />
```

## Motivation 🔦

The discussion started at https://github.com/mui-org/material-ui-x/pull/1283#discussion_r600368460. There are a number of translation files that didn't handle the aria-label:

https://github.com/mui-org/material-ui-x/blob/cad4f90b160f5847cebea9ba67c87c8786741dfa/packages/grid/_modules_/grid/locales/deDE.ts#L7
https://github.com/mui-org/material-ui-x/blob/8263b28cec3468a9803b562f6b6036b6b10d3567/packages/grid/_modules_/grid/locales/plPL.ts#L7
https://github.com/mui-org/material-ui-x/blob/69d2b6501d3b12662da7e864f6f4765be9dbc07c/packages/grid/_modules_/grid/locales/nlNL.ts#L7
https://github.com/mui-org/material-ui-x/blob/874539d49d13c350713a0a635bde5a7f213b6a7a/packages/grid/_modules_/grid/locales/jaJP.ts#L7
https://github.com/mui-org/material-ui-x/blob/4be94765e1885c205bc7407a3b74a7474f4ccd48/packages/grid/_modules_/grid/locales/frFR.ts#L7

As far as I know, adding a hardcoded generic aria-label:

- isn't helpful for accessibility screenreaders. The information is already available in `role="grid"`.
- prevents `aria-labelledby` to behave correctly if it's how developers prefer to label the component. 

We do have one hardcoded aria-label in the core for a specific action: https://github.com/mui-org/material-ui/blob/1b096070faf102281f8e3c4f9b2bf50acf91f412/packages/material-ui/src/Breadcrumbs/Breadcrumbs.js#L105

Therefore, I propose we remove the generic `aria-label="grid"` from the source and the translations. cc @eps1lon

The guidelines: https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html.

### LINKED ISSUE #1700: [DataGrid] customizable aria-label or aria-labelledby field
Current DataGrid doesn't seem to have a way to receive a user defined aria-label or aria-labelledby reference.  This is needed for accessibility.

- [x] I have searched the [issues](https://github.com/mui-org/material-ui/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Most components need a descriptive label for screen readers to identify them to users.  Current implementation of DataGrid seems to hardcode the label to "grid":

Screen Shot 2021-05-18 at 1 12 10 PM

Screen Shot 2021-05-18 at 1 13 21 PM

## Examples 🌈
Here's a devTool screenshot of this WAI aria page demonstrating accessible tables:  
https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html

Screen Shot 2021-05-18 at 1 16 45 PM

## Motivation 🔦
a11y compliance

### PR DESCRIPTION
**Pull request by ZeeshanTamboli**
Closes #1700 
Closes #1290

### CODE REVIEW

File: packages/grid/_modules_/grid/GridComponent.tsx [hunk 1]
```diff
@@ -111,12 +111,17 @@ export const GridComponent = React.forwardRef<HTMLDivElement, GridComponentProps
...
+    const ariaProps = {
+      'aria-label': props['aria-label'],
+    };
     return (
       <GridApiContext.Provider value={apiRef}>
         <NoSsr>
           <GridRoot
             ref={handleRef}
             className={clsx(internalOptions.classes?.root, props.className)}
+            {...ariaProps}
```
Comment (dtassone) [inline packages/grid/_modules_/grid/GridComponent.tsx:124]: why don't you put the prop directly?
`aria-label= {props['aria-label']}`

File: packages/grid/_modules_/grid/components/containers/GridRoot.tsx [hunk 2]
```diff
@@ -27,7 +27,6 @@ export const GridRoot = React.forwardRef<HTMLDivElement, GridRootProps>(function
       role="grid"
       aria-colcount={visibleColumnsLength}
       aria-rowcount={gridState.rows.totalRowCount}
-      aria-label={apiRef!.current.getLocaleText('rootGridLabel')}
```
Comment (dtassone) [inline packages/grid/_modules_/grid/components/containers/GridRoot.tsx:30]: Why don't we keep a default value?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/components/containers/GridRoot.tsx:30]: #1290

File: packages/grid/_modules_/grid/GridComponent.tsx [hunk 3]
```diff
@@ -111,12 +111,17 @@ export const GridComponent = React.forwardRef<HTMLDivElement, GridComponentProps
 
     const showNoRowsOverlay = !props.loading && totalRowCount === 0;
     const showNoResultsOverlay = !props.loading && totalRowCount > 0 && visibleRowCount === 0;
+
+    const ariaProps = {
+      'aria-label': props['aria-label'],
+    };
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/GridComponent.tsx:117]: AFAIK, it's also important to add:
```suggestion
      'aria-labelledby': props['aria-labelledby'],
    };
```

for instance: https://www.w3.org/TR/wai-aria-practices/examples/grid/dataGrids.html

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponent.tsx:117]: I thought about it. AFAIK, we need a corresponding `id` for `aria-labelledby` where the id refers to the element which has a text needed for labelling. Do we have `header` or any element where we show static text?

Comment (oliviertassinari) [reply packages/grid/_modules_/grid/GridComponent.tsx:117]: It's up to the developers to provide, my point is that we currently prevent them to do it.

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponent.tsx:118]: I added `aria-labelledby` on the root.

**Comment by ZeeshanTamboli** [comment]
Do we have to document anything in the [accessibility doc](https://material-ui.com/components/data-grid/accessibility/)?

**Comment by oliviertassinari** [comment]
Note sure, we don't have much either for the Table: https://next.material-ui.com/components/tables/#accessibility.

File: packages/grid/_modules_/grid/GridComponentProps.ts [hunk 4]
```diff
@@ -68,4 +68,12 @@ export interface GridComponentProps extends GridOptionsProp {
...
    */
   getRowId?: GridRowIdGetter;
+  /**
+   * The label of the grid.
+   */
+  'aria-label'?: string;
+  /**
+   * The id of the element containing a label for the grid.
+   */
+  'aria-labelledby'?: string;
```
Comment (oliviertassinari) [inline packages/grid/_modules_/grid/GridComponentProps.ts:78]: We sort the props alphabetically

Comment (ZeeshanTamboli) [reply packages/grid/_modules_/grid/GridComponentProps.ts]: Sorted all the props.

**Review by dtassone — APPROVED** [review]

**Pull request merged**